### PR TITLE
Add a design for leader to leader transition on slot boundaries.

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Leader to leader transition on slot boundaries is complex because the next leader can arrive at its slot concurrently in the middle of the current slot's block.  This design addresses those challenges.

#### Summary of Changes

Add a design and an alternative covering how the next leader should behave when it arrives at its own slot.

Fixes #

### CODE REVIEW

File: book/src/leader-leader-handoff.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+for the current slot.  The next leader may arrive at that slot while still
+processing the current leaders entries.
+
+The ideal scenario would be that the next leader generated its own slot right
+after it was able to vote for the current leader.  Otherwise it is possible that
+the next leader will produce a block that the current leader failed while the
+rest of the network observes that block succeeding.
+
+The current leader has incentives to start its slot as early as possible to
+capture economic rewards.  Those incentives need to be balanced by the leaders
```
Comment (rob-solana) [inline book/src/leader-leader-handoff.md]: leaders => leader's

File: book/src/leader-leader-handoff.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+need to attach its block to a block that has the most commitment from the rest
+of the network.
+
+## Leader timeout
+
+A leader delays its slot in real time, while it is actively receiving entries
+for the previous slot.  The delay is locally configurable by each leader, and can
+be dynamically based on the previous leaders behavior.  If the previous leaders
```
Comment (rob-solana) [inline book/src/leader-leader-handoff.md]: leaders=>leader's

**Comment by rob-solana** [comment]
there's a 3rd option: what if the next leader speculatively generated the ultimate tick and his first tick from entries received since the penultimate tick?  Next leader would be burning 2 cores toward the end of the previous leader's slot, but there could be zero delay in the case where the network latency is similar to 100ms...

**Comment by aeyakovenko** [comment]
> there's a 3rd option: what if the next leader speculatively generated the ultimate tick and his first tick from the most recently observed entry received since the penultimate tick? Next leader would be burning 2 cores toward the end of the previous leader's slot, but there could be zero delay in the case where the network latency is similar to 100ms...
> 
> Obviously we still need the timeout, but we wouldn't need a guard tick.

I think this optimization is available to the guard based approach.  But without a guard the previous leader can insert a transaction simply to mess with the next one after the *penultimate* tick.  I think it would be hard to defend against.

File: book/src/SUMMARY.md [hunk 3]
```diff
@@ -36,6 +36,7 @@
   - [Data Plane Fanout](data-plane-fanout.md)
   - [Reliable Vote Transmission](reliable-vote-transmission.md)
   - [Bank Forks](bank-forks.md)
+  - [Leader to Leader Handoff](leader-leader-handoff.md)
```
Comment (garious) [inline book/src/SUMMARY.md]: Your other proposal had a very similar title "Leader to Validator Transition" with filename "leader-validator-transition.md". Here you use the term "handoff". Why the difference?

Comment (aeyakovenko) [reply book/src/SUMMARY.md]: @garious I could use the same term.  The leader to validator transition was the transition inside the fullnode.  This transition is covering how each leader hands-off PoH production to the next one.

File: book/src/leader-leader-handoff.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
+# Leader to Leader Handoff
+
+This design describes how leader's handoff production of the PoH ledger between
```
Comment (rob-solana) [inline book/src/leader-leader-handoff.md]: leaders

File: book/src/leader-leader-handoff.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+* All the space in a block is used for entries.
+
+* The timeout is not fixed.
+
+* The timeout is local to the leader, and therefore can be clever.  The leader's
+heuristic can take into account avalanche performance.
+
+* This design doesn't require a ledger hard fork to update.
+
+* The previous leader can redundantly transmit the last entry in the block to
```
Comment (rob-solana) [inline book/src/leader-leader-handoff.md]: what if we changed Broadcast/Avalanche to send every Blob to the next leader (as well as spreading it around the network)?

Comment (aeyakovenko) [reply book/src/leader-leader-handoff.md]: @rob-solana,  double the bandwidth. 

The next leader should only pick the previous block if the rest of the network will pick it as well.  Actually seeing the retransmission of the blobs is a pretty good signal of how far the packets are getting.

File: book/src/SUMMARY.md (hunk 3 above)
Comment (garious) [reply book/src/SUMMARY.md]: Doesn't matter. Just use the same word consistently.

**Pull request merged**